experimental/cuda2/dynamic_symbols.c

```diff
@@ -0,0 +1,228 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "experimental/cuda2/dynamic_symbols.h"
+
+#include <string.h>
+
+#include "experimental/cuda2/status_util.h"
+#include "iree/base/assert.h"
+#include "iree/base/internal/dynamic_library.h"
+#include "iree/base/target_platform.h"
+#include "iree/base/tracing.h"
+
+//===----------------------------------------------------------------------===//
+// CUDA dynamic symbols
+//===----------------------------------------------------------------------===//
+
+static const char* iree_hal_cuda_dylib_names[] = {
+#if defined(IREE_PLATFORM_WINDOWS)
+    "nvcuda.dll",
+#else
+    "libcuda.so",
+#endif  // IREE_PLATFORM_WINDOWS
+};
+
+#define IREE_CONCAT(A, B) A B
+
+// Resolves all CUDA dynamic symbols in `dynamic_symbol_tables.h`, prefer _v2
+// version if it exists.
+static iree_status_t iree_hal_cuda2_dynamic_symbols_resolve_all(
+    iree_hal_cuda2_dynamic_symbols_t* syms) {
+#define IREE_CU_PFN_DECL(cuda_symbol_name, ...)                         \
+  {                                                                     \
+    static const char* name = #cuda_symbol_name;                        \
+    IREE_RETURN_IF_ERROR(iree_dynamic_library_lookup_symbol(            \
+        syms->dylib, name, (void**)&syms->cuda_symbol_name));           \
+    static const char* name_v2 = IREE_CONCAT(#cuda_symbol_name, "_v2"); \
+    void* fptr_v2;                                                      \
+    iree_dynamic_library_lookup_symbol(syms->dylib, name_v2, &fptr_v2); \
+    if (fptr_v2) syms->cuda_symbol_name = fptr_v2;                      \
+  }
+#define IREE_NCCL_PFN_DECL(nccl_symbol_name, ...)  // Ignore NCCL symbols
```

Comment:
yeah, splitting out the headers in a follow-up would make this much cleaner

Reply:
Yes. Here it is: https://github.com/openxla/iree/pull/13954